include/valijson/validation_visitor.hpp

```diff
@@ -347,6 +347,82 @@ class ValidationVisitor: public constraints::ConstraintVisitor
         return numValidated > 0;
     }
 
+    /**
+    * @brief    Validate current node against a FormatConstraint
+    * 
+    * @param    constraint  Constraint that the target must validate against
+    * 
+    * @return   \c true if validation succeeds; \c false otherwise
+    */
+    bool visit(const FormatConstraint &constraint) override
+    {
+        const std::string s = m_target.asString();
+        const std::string format = constraint.getFormat();
+        if (format == "date")
+        {
+            std::regex date_regex("^([0-9]+)-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])$");
+            std::smatch matches;
+            if (std::regex_match(s, matches, date_regex)) {
+                const auto month = std::stoi(matches[2].str());
+                const auto day = std::stoi(matches[3].str());
+                if (month == 2) {
+                    if (day > 29) {
+                        if (m_results) {
+                            m_results->pushError(m_context,
+                                "String should be a valid date");
+                        }
+                        return false;
+                    }
+                }
+                return true;
+            } else {
+                if (m_results) {
+                    m_results->pushError(m_context,
+                                         "String should be a valid date");
+                }
+                return false;
+            }
+        } else if (format == "time") {
+            std::regex time_regex("^([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(\.[0-9]+)?(([Zz])|([\+|\-]([01][0-9]|2[0-3]):[0-5][0-9]))$");
+            if (std::regex_match(s, time_regex)) {
+                return true;
+            } else {
+                if (m_results) {
+                    m_results->pushError(m_context,
+                                         "String should be a valid time");
+                }
+                return false;
+            }
+        } else if (format == "date-time") {
+            std::regex datetime_regex("^([0-9]+)-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])[Tt]([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(\.[0-9]+)?(([Zz])|([\+|\-]([01][0-9]|2[0-3]):[0-5][0-9]))$");
+            std::smatch matches;
+            if (std::regex_match(s, matches, datetime_regex)) {
+                //const auto year = std::stoi(matches[1].str());
+                const auto month = std::stoi(matches[2].str());
+                const auto day = std::stoi(matches[3].str());
+                if (month == 2) {
+                    if (day > 29) {
+                        if (m_results) {
```

Comment:
Do you have a specific use-case that benefitted from validating this particular invariant about date strings? My thinking is that an all-or-nothing approach would lead to more consistent behaviour here. e.g. if "2022/12/32" validates but "2022/02/30" does not, that would seem inconsistent to me.

So maybe it would be better to leave this to the application to handle?

Reply:
`            {
                "description": "a invalid day in date-time string",
                "data": "1990-02-31T15:59:60.123-08:00",
                "valid": false
            },

This was added to pass the above test case from thirdparty\JSON-Schema-Test-Suite\tests\draft7\optional\format\date-time.json

A day bigger than 31 should not pass the regular expression.

In general I agree that the precision of validation here is a bit arbitrary. At least for our use cases just verifying the syntax and not the semantics would be sufficient.

Reply:
Makes sense. In this case, I would consider moving the fine-grained validation to a helper method. That would reduce some duplication. I would also consider making this check more complete. i.e. Checking the day-range for other months, not just February. What are your thoughts on doing this?